File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();

        map.put(0, 1);
        int ans = 0;
        int prefix_sum = 0;

        for (int ele : nums) {
            prefix_sum += ele;
            int find = prefix_sum - k;

            if (map.containsKey(find)) {
                int freq = map.get(find);
                ans += freq;
            }

            if (map.containsKey(prefix_sum))
                map.put(prefix_sum, map.get(prefix_sum) + 1);
            else
                map.put(prefix_sum, 1);
        }

        return ans;
    }
}

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution(object):
    def subarraySum(self, nums, k):
        
        map = {}
        map[0] = 1
        
        prefix_sum = 0
        ans = 0

        for ele in nums:

            prefix_sum += ele
            find = prefix_sum - k
            
            if find in map:
                ans += map[find]

            if prefix_sum in map:
                map[prefix_sum] += 1
            else:
                map[prefix_sum] = 1
        
        return ans